File: XTApp/XTApp/XTLogin/XTLoginCodeVC.m
//
//  XTLoginCodeVC.m
//  XTApp
//
//  Created by xia on 2024/9/4.
//

#import "XTLoginCodeVC.h"
#import "XTPhoneCodeApi.h"
#import "XTLoginVC.h"
#import "XTHtmlVC.h"
#import "XTCodeAltView.h"
#import <YFPopView/YFPopView.h>
#import "XTLoginVC.h"

@interface XTLoginCodeVC ()

@property(nonatomic,copy) NSString *countDown;
@property(nonatomic,copy) NSString *phone;
@property(nonatomic,weak) XTLoginVC *loginVC;
@property(nonatomic,weak) UITextField *textField;
@property(nonatomic,weak) dispatch_source_t codeTimer;

@end

@implementation XTLoginCodeVC

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    UIImageView *iconImg = [UIImageView xt_img:@"xt_login_icon" tag:0];
    [self.xt_navView addSubview:iconImg];
    [iconImg mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(self.xt_navView.mas_left).offset(20);
        make.centerY.equalTo(self.xt_bkBtn);
    }];
    
    UILabel *appNameLab = [UILabel xt_lab:CGRectZero text:XT_App_Name font:XT_Font_SD(16) textColor:[UIColor whiteColor] alignment:NSTextAlignmentLeft tag:0];
    [self.xt_navView addSubview:appNameLab];
    [appNameLab mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(iconImg.mas_right).offset(10);
        make.centerY.equalTo(self.xt_bkBtn);
    }];
    
    self.xt_bkBtn.hidden = YES;
    UIImageView *bgImg = [UIImageView xt_img:@"xt_login_code_top_bg" tag:0];
    [self.view addSubview:bgImg];
    [bgImg mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.right.equalTo(self.view);
        make.top.equalTo(self.view.mas_top);
    }];
    self.xt_navView.backgroundColor = [UIColor clearColor];
    [self.view bringSubviewToFront:self.xt_navView];
    
    UIView *view = [UIView new];
    view.clipsToBounds = YES;
    view.layer.cornerRadius = 20;
    [self.view addSubview:view];
    [view mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.right.equalTo(self.view);
        make.top.equalTo(bgImg.mas_bottom).offset(-23);
        make.height.mas_equalTo(305);
    }];
    
    [view.layer addSublayer:[UIView xt_layer:@[(__bridge id)XT_RGB(0xF3FF9B, 1.0f).CGColor,(__bridge id)XT_RGB(0xFFFFFF, 1.0f).CGColor] locations:@[@0,@1.0f] startPoint:CGPointMake(0.5, 0) endPoint:CGPointMake(0.5, 0.48) size:CGSizeMake(self.view.width, 305)]];
    
    UILabel *titLab = [UILabel xt_lab:@"" font:XT_Font(17) textColor:[UIColor blackColor] alignment:NSTextAlignmentLeft isPriority:YES tag:0];
    titLab.numberOfLines = 0;
    [self.view addSubview:titLab];
    [titLab mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(self.view.mas_left).offset(27);
        make.right.equalTo(self.view.mas_right).offset(-27);
        make.top.equalTo(view.mas_top).offset(23);
    }];
    
    titLab.attributedText = [NSString xt_strs:@[@"Log in",@"to experience personal loan services now"] fonts:@[XT_Font_M(31),XT_Font_M(17)] colors:@[XT_RGB(0x0BB559, 1.0f),XT_RGB(0x000000, 1.0f)]];
    
    
    UIView *codeView = [UIView xt_frame:CGRectZero color:[UIColor whiteColor]];
    codeView.layer.cornerRadius = 26;
    codeView.layer.shadowColor = [UIColor colorWithRed:0/255.0 green:0/255.0 blue:0/255.0 alpha:0.1100].CGColor;
    codeView.layer.shadowOffset = CGSizeMake(0,1);
    codeView.layer.shadowOpacity = 1;
    codeView.layer.shadowRadius = 4;
    [self.view addSubview:codeView];
    [codeView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(self.view.mas_left).offset(20);
        make.right.equalTo(self.view.mas_right).offset(-20);
        make.top.equalTo(titLab.mas_bottom).offset(20);
        make.height.mas_equalTo(52);
    }];
    
    UILabel *logoLab = [UILabel xt_lab:@"+63" font:XT_Font_M(17) textColor:XT_RGB(0x0BB559, 1.0f) alignment:NSTextAlignmentCenter isPriority:YES tag:0];
    [self.view addSubview:logoLab];
    [logoLab mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(codeView.mas_left).offset(26);
        make.centerY.equalTo(codeView);
        make.height.mas_equalTo(20);
    }];
    
    UITextField *phoneTextField = [UITextField xt_textField:NO placeholder:@"Phone number" font:XT_Font_M(14) textColor:[UIColor blackColor] withdelegate:self];
    phoneTextField.keyboardType = UIKeyboardTypeNumberPad;
    [self.view addSubview:phoneTextField];
    self.textField = phoneTextField;
    [phoneTextField addTarget:self action:@selector(textFieldDidChange:) forControlEvents:UIControlEventEditingChanged];
    [phoneTextField mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(logoLab.mas_right).offset(27);
        make.right.equalTo(codeView.mas_right).offset(-20);
        make.centerY.equalTo(codeView);
        make.height.mas_equalTo(codeView);
    }];
    
    UIButton *readBtn = [UIButton xt_btn:@" I have read and agree with the" font:XT_Font(12) textColor:[UIColor blackColor] cornerRadius:0 tag:0];
    [self.view addSubview:readBtn];
    [readBtn setImage:XT_Img(@"xt_login_code_select_yes") forState:UIControlStateSelected];
    [readBtn setImage:XT_Img(@"xt_login_code_select_no") forState:UIControlStateNormal];
    [readBtn mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(codeView.mas_left).offset(16);
        make.top.equalTo(phoneTextField.mas_bottom).offset(25);
        make.height.mas_equalTo(25);
    }];
    readBtn.selected = YES;
    readBtn.rac_command = [[RACCommand alloc]initWithSignalBlock:^RACSignal *(id input) {
        readBtn.selected = !readBtn.selected;
        return [RACSignal empty];
    }];
    
    UIButton *privacyBtn = [UIButton xt_btn:@"" font:nil textColor:nil cornerRadius:0 tag:0];
    NSAttributedString *str = [[NSAttributedString alloc] initWithString:@"Privacy Agreement" attributes:@{
        NSFontAttributeName:XT_Font_B(12),
        NSForegroundColorAttributeName:XT_RGB(0x02CC56, 1.0f),
        NSUnderlineStyleAttributeName:[NSString stringWithFormat:@"%ld", NSUnderlineStyleSingle],
    }];
    [privacyBtn setAttributedTitle:str forState:UIControlStateNormal];
    [self.view addSubview:privacyBtn];
    [privacyBtn mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(readBtn.mas_right);
        make.centerY.equalTo(readBtn);
        make.height.mas_equalTo(readBtn);
    }];
    @weakify(self)
    privacyBtn.rac_command = [[RACCommand alloc]initWithSignalBlock:^RACSignal *(id input) {
        @strongify(self)
        XTHtmlVC *htmlVC = [[XTHtmlVC alloc] initUrl:XT_Privacy_Url];
        [self.navigationController pushViewController:htmlVC animated:YES];
        return [RACSignal empty];
    }];
    
    UIButton *codeBtn = [UIButton xt_btn:@"Let’s Go" font:XT_Font_B(20) textColor:[UIColor whiteColor] cornerRadius:24 tag:0];
    codeBtn.backgroundColor = XT_RGB(0x02CC56, 1.0f);
    [self.view addSubview:codeBtn];
    [codeBtn mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(self.view.mas_left).offset(20);
        make.right.equalTo(self.view.mas_right).offset(-20);
        make.top.equalTo(readBtn.mas_bottom).offset(30);
        make.height.mas_equalTo(48);
    }];
    codeBtn.rac_command = [[RACCommand alloc]initWithSignalBlock:^RACSignal *(id input) {
        @strongify(self)
//        self.countDown = @"60";
//        [self nextLoginVC];
//        [self goCountDown];
//        return [RACSignal empty];
        if(!readBtn.selected) {
            [self showAltView];
            return [RACSignal empty];
        }
        [self checkCode];
        return [RACSignal empty];
    }];
    
    UIButton *testBtn = [UIButton new];
    [self.view addSubview:testBtn];
    [testBtn mas_makeConstraints:^(MASConstraintMaker *make) {
        make.right.equalTo(self.view.mas_right);
        make.bottom.equalTo(self.view.mas_bottom);
        make.size.mas_equalTo(CGSizeMake(50, 50 + XT_Bottom_Height));
    }];
    __block NSInteger tap = 0;
    testBtn.rac_command = [[RACCommand alloc]initWithSignalBlock:^RACSignal *(id input) {
        @strongify(self)
        tap ++;
        if(tap == 5) {
            tap = 0;
            [self xt_altView];
        }
        
        return [RACSignal empty];
    }];
}

-(void)xt_altView {
    UIAlertController *altVC = [UIAlertController alertControllerWithTitle:@"switch_environment" message:nil preferredStyle:UIAlertControllerStyleAlert];
    
    [altVC addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.placeholder = @"switch_environment";
    }];
    UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
    }];
    
    [altVC addAction:cancelAction];
    @weakify(self)
    UIAlertAction *sureAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDestructive handler:^(UIAlertAction * _Nonnull action) {
        @strongify(self)
        UITextField *textField = altVC.textFields.firstObject;
        if(![textField.text isEqualToString:@"ksm2023"]) {
            [XTUtility xt_showTips:@"Password error！" view:nil];
            return;
        }
        [self xt_change];
    }];
    
    [altVC addAction:sureAction];
    
    [self xt_presentViewController:altVC animated:YES completion:nil modalPresentationStyle:UIModalPresentationFullScreen];
    
}

-(void)xt_change {
    UIAlertController *altVC = [UIAlertController alertControllerWithTitle:@"Environment" message:nil preferredStyle:UIAlertControllerStyleAlert];
    
    [altVC addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.placeholder = @"Environment";
        textField.keyboardType = UIKeyboardTypeURL;
    }];
    UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
    }];
    
    [altVC addAction:cancelAction];
    UIAlertAction *sureAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDestructive handler:^(UIAlertAction * _Nonnull action) {
        UITextField *textField = altVC.textFields.firstObject;
        if(![textField.text hasPrefix:@"http://"] && ![textField.text hasPrefix:@"https://"]) {
            [XTUtility xt_showTips:@"Please enter the correct domain name" view:nil];
            return;
        }
        NSString *url = [NSString stringWithFormat:@"%@%@",textField.text,XT_Api_api];
        [url writeToFile:XT_Locality_Url_Path atomically:YES encoding:NSUTF8StringEncoding error:nil];

    }];
    
    [altVC addAction:sureAction];
    
    [self xt_presentViewController:altVC animated:YES completion:nil modalPresentationStyle:UIModalPresentationFullScreen];
}

- (void)textFieldDidChange:(UITextField *)textField {
    if (textField == self.textField && textField.text.length > 15) {
        self.textField.text =  [self.textField.text substringToIndex:15];
    }
}

- (void)showAltView {
    XTCodeAltView *view = [[XTCodeAltView alloc] init];
    view.center = self.view.center;
    YFPopView *popView = [[YFPopView alloc] initWithAnimationView:view];
    popView.animationStyle = YFPopViewAnimationStyleFade;
    popView.autoRemoveEnable = YES;
    [popView showPopViewOn:self.view];
}

- (void)checkCode {
    NSString *text = XT_Object_To_Stirng(self.textField.text);
    if(text.length > 15 || text.length < 8){
        [XTUtility xt_showTips:@"Please enter a valid phone number" view:self.view];
        return;
    }
    if([self.phone isEqualToString:text] && [self.countDown integerValue] > 0){
        [self nextLoginVC];
        return;
    }
    self.phone = text;
    @weakify(self)
    [self getCodeNumber:^{
        @strongify(self)
        [self nextLoginVC];
    }];
}

-(void)getCodeNumber:(XTBlock)block {
    @weakify(self)
    XTPhoneCodeApi *api = [[XTPhoneCodeApi alloc] initPhone:self.phone];
    [XTUtility xt_showProgress:self.view message:@"loading..."];
    [api xt_startRequestSuccess:^(NSDictionary *dic, NSString *str) {
        @strongify(self)
        [XTUtility xt_atHideProgress:self.view];
        [XTUtility xt_showTips:str view:nil];
        if([dic isKindOfClass:[NSDictionary class]] && [dic[@"gugosixyleNc"] isKindOfClass:[NSDictionary class]]) {
            self.countDown = XT_Object_To_Stirng(dic[@"gugosixyleNc"][@"tedisixurnalNc"]);
            [self goCountDown];
            if(block) {
                block();
            }
        }
    } failure:^(NSDictionary *dic, NSString *str) {
        @strongify(self)
        [XTUtility xt_atHideProgress:self.view];
        [XTUtility xt_showTips:str view:nil];
    } error:^(NSError * _Nonnull error) {
        @strongify(self)
        [XTUtility xt_atHideProgress:self.view];
    }];
}

-(void)goCountDown {
    if(self.codeTimer){
        dispatch_source_cancel(self.codeTimer);
    }
    __weak XTLoginCodeVC *weakSelf = self;
    __block NSInteger timeout = [self.countDown integerValue]; //倒计时时间
    dispatch_queue_t queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
    dispatch_source_t timer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0,queue);
    self.codeTimer = timer;
    dispatch_source_set_timer(timer,dispatch_walltime(NULL, 0),1.0*NSEC_PER_SEC, 0); //每秒执行
    dispatch_source_set_event_handler(timer, ^{
        if(!weakSelf) {
            dispatch_source_cancel(timer);
        }
        if(timeout<=0){ //倒计时结束，关闭
            dispatch_source_cancel(timer);
            weakSelf.countDown = @"0";
        }
        else {
            timeout--;
            weakSelf.countDown = [NSString stringWithFormat:@"%ld",timeout];
        }
        if(timeout<=0){ //倒计时结束，关闭
            dispatch_source_cancel(timer);
            weakSelf.countDown = @"0";
        }
        if(weakSelf.loginVC) {
            dispatch_async(dispatch_get_main_queue(), ^{
                [weakSelf.loginVC reloadCountDown:weakSelf.countDown];
            });
        }
    });
    dispatch_resume(timer);
}

-(void)nextLoginVC {
    XTLoginVC *vc = [[XTLoginVC alloc] initPhone:self.phone countDown:self.countDown];
    [self.navigationController pushViewController:vc animated:YES];
    vc.loginBlock = self.loginBlock;
    self.loginVC = vc;
    @weakify(self)
    vc.resendBlock = ^{
        @strongify(self)
        [self getCodeNumber:nil];
    };
    
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: XTApp/XTApp/XTLogin/XTLoginCodeVC.swift
//
//  XTLoginCodeVC.swift
//  XTApp
//
//  Created by Codex on 2026/4/30.
//

import UIKit
import YFPopView

@objcMembers
@objc(XTLoginCodeVC)
class XTLoginCodeVC: XTBaseVC, UITextFieldDelegate {
    dynamic var loginBlock: XTBlock?

    private var countDown = "0"
    private var phone = ""
    private weak var loginVC: XTLoginVC?
    private weak var textField: UITextField?
    private weak var agreementButton: UIButton?
    private var codeTimer: DispatchSourceTimer?
    private var environmentTapCount = 0

    override func viewDidLoad() {
        super.viewDidLoad()
        xt_bkBtn.isHidden = true
        xt_navView.backgroundColor = .clear
        setupHeader()
        setupForm()
    }

    deinit {
        codeTimer?.cancel()
    }

    private func setupHeader() {
        let icon = UIImageView(image: XT_Img("xt_login_icon"))
        icon.frame = CGRect(x: 20, y: XT_StatusBar_Height + 8, width: 28, height: 28)
        xt_navView.addSubview(icon)

        let name = UILabel(frame: CGRect(x: 58, y: icon.y, width: 180, height: 28))
        name.text = XT_App_Name
        name.font = XT_Font_SD(16)
        name.textColor = .white
        xt_navView.addSubview(name)

        let top = UIImageView(image: XT_Img("xt_login_code_top_bg"))
        top.frame = CGRect(x: 0, y: 0, width: view.width, height: 220)
        top.contentMode = .scaleAspectFill
        view.insertSubview(top, belowSubview: xt_navView)
    }

    private func setupForm() {
        let panel = UIView(frame: CGRect(x: 0, y: 190, width: view.width, height: 305))
        panel.backgroundColor = .white
        panel.layer.cornerRadius = 20
        panel.clipsToBounds = true
        view.addSubview(panel)
        view.bringSubviewToFront(xt_navView)

        let title = UILabel(frame: CGRect(x: 27, y: 23, width: view.width - 54, height: 72))
        title.numberOfLines = 0
        title.attributedText = NSString.xt_strs(
            ["Log in\n", "to experience personal loan services now"],
            fonts: [XT_Font_M(31), XT_Font_M(17)],
            colors: [XT_RGB(0x0BB559, 1.0), .black]
        )
        panel.addSubview(title)

        let codeView = UIView(frame: CGRect(x: 20, y: 116, width: panel.width - 40, height: 52))
        codeView.backgroundColor = .white
        codeView.layer.cornerRadius = 26
        codeView.layer.shadowColor = UIColor.black.withAlphaComponent(0.11).cgColor
        codeView.layer.shadowOffset = CGSize(width: 0, height: 1)
        codeView.layer.shadowOpacity = 1
        codeView.layer.shadowRadius = 4
        panel.addSubview(codeView)

        let prefix = UILabel(frame: CGRect(x: 26, y: 0, width: 42, height: 52))
        prefix.text = "+63"
        prefix.font = XT_Font_M(17)
        prefix.textColor = XT_RGB(0x0BB559, 1.0)
        codeView.addSubview(prefix)

        let phoneField = UITextField(frame: CGRect(x: 92, y: 0, width: codeView.width - 112, height: 52))
        phoneField.placeholder = "Phone number"
        phoneField.font = XT_Font_M(14)
        phoneField.keyboardType = .numberPad
        phoneField.delegate = self
        phoneField.addTarget(self, action: #selector(textFieldDidChange(_:)), for: .editingChanged)
        codeView.addSubview(phoneField)
        textField = phoneField

        let readBtn = UIButton(type: .custom)
        readBtn.frame = CGRect(x: 36, y: 193, width: 210, height: 25)
        readBtn.setTitle(" I have read and agree with the", for: .normal)
        readBtn.setTitleColor(.black, for: .normal)
        readBtn.titleLabel?.font = XT_Font(12)
        readBtn.setImage(XT_Img("xt_login_code_select_yes"), for: .selected)
        readBtn.setImage(XT_Img("xt_login_code_select_no"), for: .normal)
        readBtn.isSelected = true
        readBtn.addTarget(self, action: #selector(toggleRead(_:)), for: .touchUpInside)
        panel.addSubview(readBtn)
        agreementButton = readBtn

        let privacy = UIButton(type: .custom)
        privacy.frame = CGRect(x: readBtn.frame.maxX, y: 193, width: 130, height: 25)
        privacy.setAttributedTitle(NSAttributedString(
            string: "Privacy Agreement",
            attributes: [
                .font: XT_Font_B(12),
                .foregroundColor: XT_RGB(0x02CC56, 1.0),
                .underlineStyle: NSUnderlineStyle.single.rawValue
            ]
        ), for: .normal)
        privacy.addTarget(self, action: #selector(openPrivacy), for: .touchUpInside)
        panel.addSubview(privacy)

        let codeBtn = UIButton.xt_btn("Let’s Go", font: XT_Font_B(20), textColor: .white, cornerRadius: 24, tag: 0)
        codeBtn.backgroundColor = XT_RGB(0x02CC56, 1.0)
        codeBtn.frame = CGRect(x: 20, y: 248, width: panel.width - 40, height: 48)
        codeBtn.addTarget(self, action: #selector(letGoTapped), for: .touchUpInside)
        panel.addSubview(codeBtn)

        let testBtn = UIButton(type: .custom)
        testBtn.frame = CGRect(x: view.width - 50, y: view.height - 50 - XT_Bottom_Height, width: 50, height: 50 + XT_Bottom_Height)
        testBtn.addTarget(self, action: #selector(environmentButtonTapped), for: .touchUpInside)
        view.addSubview(testBtn)
    }

    @objc private func letGoTapped() {
        guard agreementButton?.isSelected == true else {
            showAltView()
            return
        }
        checkCode()
    }

    @objc private func toggleRead(_ sender: UIButton) {
        sender.isSelected.toggle()
    }

    @objc private func openPrivacy() {
        navigationController?.pushViewController(XTHtmlVC(url: XT_Privacy_Url), animated: true)
    }

    @objc func textFieldDidChange(_ textField: UITextField) {
        if textField.text?.count ?? 0 > 15 {
            textField.text = String(textField.text?.prefix(15) ?? "")
        }
    }

    private func showAltView() {
        let altView = XTCodeAltView()
        altView.center = view.center
        guard let popView = YFPopView(animationView: altView) else { return }
        popView.animationStyle = .fade
        popView.autoRemoveEnable = true
        popView.show(on: view)
    }

    @objc private func environmentButtonTapped() {
        environmentTapCount += 1
        guard environmentTapCount == 5 else { return }
        environmentTapCount = 0
        showEnvironmentPasswordAlert()
    }

    private func showEnvironmentPasswordAlert() {
        let alert = UIAlertController(title: "switch_environment", message: nil, preferredStyle: .alert)
        alert.addTextField { textField in
            textField.placeholder = "switch_environment"
        }
        alert.addAction(UIAlertAction(title: "Cancel", style: .cancel))
        alert.addAction(UIAlertAction(title: "OK", style: .destructive) { [weak self, weak alert] _ in
            guard alert?.textFields?.first?.text == "ksm2023" else {
                XTUtility.xt_showTips("Password error!", view: nil)
                return
            }
            self?.showEnvironmentInputAlert()
        })
        xt_presentViewController(alert, animated: true, completion: nil, modalPresentationStyle: .fullScreen)
    }

    private func showEnvironmentInputAlert() {
        let alert = UIAlertController(title: "Environment", message: nil, preferredStyle: .alert)
        alert.addTextField { textField in
            textField.placeholder = "Environment"
            textField.keyboardType = .URL
        }
        alert.addAction(UIAlertAction(title: "Cancel", style: .cancel))
        alert.addAction(UIAlertAction(title: "OK", style: .destructive) { [weak alert] _ in
            let text = alert?.textFields?.first?.text ?? ""
            guard text.hasPrefix("http://") || text.hasPrefix("https://") else {
                XTUtility.xt_showTips("Please enter the correct domain name", view: nil)
                return
            }
            let url = text + XT_Api_api
            try? url.write(toFile: XT_Locality_Url_Path, atomically: true, encoding: .utf8)
        })
        xt_presentViewController(alert, animated: true, completion: nil, modalPresentationStyle: .fullScreen)
    }

    @objc func checkCode() {
        let text = XT_Object_To_Stirng(textField?.text)
        guard (8...15).contains(text.count) else {
            XTUtility.xt_showTips("Please enter a valid phone number", view: view)
            return
        }
        if phone == text && (Int(countDown) ?? 0) > 0 {
            nextLoginVC()
            return
        }
        phone = text
        getCodeNumber { [weak self] in
            self?.nextLoginVC()
        }
    }

    @objc(getCodeNumber:)
    func getCodeNumber(_ block: XTBlock?) {
        XTLoginFlowCoordinator.shared.requestCode(phone: phone, loadingView: view, success: { [weak self] countDown in
            guard let self else { return }
            self.countDown = countDown
            self.goCountDown()
            block?()
        }, failure: nil)
    }

    private func goCountDown() {
        codeTimer?.cancel()
        var timeout = Int(countDown) ?? 0
        let timer = DispatchSource.makeTimerSource(queue: .global())
        codeTimer = timer
        timer.schedule(deadline: .now(), repeating: 1)
        timer.setEventHandler { [weak self] in
            guard let self else {
                timer.cancel()
                return
            }
            if timeout <= 0 {
                timer.cancel()
                self.countDown = "0"
            } else {
                timeout -= 1
                self.countDown = "\(timeout)"
            }
            DispatchQueue.main.async {
                self.loginVC?.reloadCountDown(self.countDown)
            }
        }
        timer.resume()
    }

    private func nextLoginVC() {
        loginVC = XTLoginFlowCoordinator.shared.pushVerification(phone: phone, countDown: countDown, from: self, loginBlock: loginBlock) { [weak self] in
            self?.getCodeNumber(nil)
        }
    }
}